rule all:
    input:
        "isolates_diversity/tree_plot.svg",
        "isolates_diversity/rooted_phylogeny/tree_plot.svg",
        "isolates_diversity/orthogroups/cds1.faa"

rule join_sfmnpv_genomes:
    input:
        "data/sfmnpv_genomes_ncbi.fna",
        "genome_assembly/sfmnpv.fa"
    output:
        "isolates_diversity/sfmnpv_genomes.fna"
    shell:
        """
        cat {input} > {output}
        """

rule align_sfmnpv_genomes:
    input:
        "isolates_diversity/sfmnpv_genomes.fna"
    output:
        "isolates_diversity/sfmnpv_genomes_aln.fna"
    shell:
        """
        mafft {input} >> {output}
        """

rule build_sfmnpv_phylogeny:
    input:
        "isolates_diversity/sfmnpv_genomes_aln.fna"
    output:
        "isolates_diversity/sfmnpv_genomes.treefile"
    params:
        "isolates_diversity/sfmnpv_genomes"
    shell:
        """
        iqtree -s {input} -pre {params} -bb 1000 -redo
        """

rule plot_sfmnpv_phylogeny:
    input:
        "isolates_diversity/sfmnpv_genomes.treefile"
    output:
        "isolates_diversity/tree_plot.svg"
    params:
        "KF891883.1"
    shell:
        """
        python src/draw_tree.py {input} {params} &&\
        mv tree_plot.svg {output}
        """

rule add_root_genomes_alignment:
    input:
        root="data/semnpv_root.fna",
        msa="isolates_diversity/sfmnpv_genomes_aln.fna"
    output:
        "isolates_diversity/rooted_phylogeny/sequences_aln.fna"
    shell:
        """
        mafft --add {input.root} {input.msa} >> {output}
        """

rule build_rooted_phylogeny:
    input:
        "isolates_diversity/rooted_phylogeny/sequences_aln.fna"
    output:
        "isolates_diversity/rooted_phylogeny/rooted_phylogeny.treefile"
    params:
        "isolates_diversity/rooted_phylogeny/rooted_phylogeny"
    shell:
        """
        iqtree -s {input} -o NC_002169.1 -pre {params} -bb 1000 -redo
        """

rule plot_rooted_phylogeny:
    input:
        "isolates_diversity/rooted_phylogeny/rooted_phylogeny.treefile"
    output:
        "isolates_diversity/rooted_phylogeny/tree_plot.svg"
    params:
        "NC_002169.1"
    shell:
        """
        python src/draw_tree.py {input} {params} && \
        mv tree_plot.svg {output}
        """

rule extract_cds:
    input:
        "genome_assembly/sfmnpv.fa",
        "annotation/sfmnpv_annotation.gtf"
    output:
        "isolates_diversity/sfmnpv_argentina_proteome.faa",
    shell:
        """
        python src/extract_protein_cds.py {input} > {output}
        """

rule cluster_orthologous_proteins:
    input:
        "isolates_diversity/sfmnpv_argentina_proteome.faa",
        "data/sfmnpv_isolates_proteomes.faa"
    output:
        "isolates_diversity/proteomes_blast.xml"
    params:
        "isolates_diversity/proteomes_db.faa"
    shell:
        """
        cat {input} > {params} && \
        makeblastdb -in {params} -dbtype prot && \
        blastp -query {input[0]} -db {params} -evalue 0.0001 -outfmt 5 -max_target_seqs 6 > {output}
        """

rule extract_orthologous_sequences:
    input:
        "isolates_diversity/proteomes_blast.xml",
        "isolates_diversity/proteomes_db.faa"
    output:
        "isolates_diversity/orthogroups/cds1.faa"
    shell:
        """
        mkdir -p isolates_diversity/orthogroups && \
        python src/get_orthogroups.py {input} && \
        mv cds* isolates_diversity/orthogroups
        """

rule align_orthologous_sequences:
    input:
        "isolates_diversity/proteomes_blast.xml",
        "isolates_diversity/proteomes_db.faa"
    output:
        "isolates_diversity/orthogroups/cds1.faa"
    shell:
        """
        mkdir -p isolates_diversity/orthogroups && \
        python src/get_orthogroups.py {input} && \
        mv cds* isolates_diversity/orthogroups
        """
#rule plot_sequence_similarity:
#    input:
#        "isolates_diversity/sfmnpv_genomes_aln.fna"
#    output:
#        "isolates_diversity/sequence_similarity_plot.svg"
#    shell:
#        """
#        python src/sequence_similarity_profile.py {input} 200 2000 &&\
#        mv sequence_similarity_plot.svg {output}
#        """
#
#rule plot_isolates_snv_distribution:
#    input:
#        "isolates_diversity/sfmnpv_genomes_snv.vcf"
#    output:
#        "isolates_diversity/isolates_snv_distribution.svg"
#    shell:
#        """
#        python src/isolates_variants_distribution.py {input} && \
#        mv isolates_snv_distribution.svg {output}
#        """
